README.txt
NSSM: The Non-Sucking Service Manager
Version 2.17-dev, 2013-08-17

NSSM is a service helper program similar to srvany and cygrunsrv.  It can 
start any application as an NT service and will restart the service if it 
fails for any reason.

NSSM also has a graphical service installer and remover.

Full documentation can be found online at

                              http://nssm.cc/

Since version 2.0, the GUI can be bypassed by entering all appropriate 
options on the command line.

Since version 2.1, NSSM can be compiled for x64 platforms.
Thanks Benjamin Mayrargue.

Since version 2.2, NSSM can be configured to take different actions
based on the exit code of the managed application.

Since version 2.3, NSSM logs to the Windows event log more elegantly.

Since version 2.5, NSSM respects environment variables in its parameters.

Since version 2.8, NSSM tries harder to shut down the managed application
gracefully and throttles restart attempts if the application doesn't run
for a minimum amount of time.

Since version 2.11, NSSM respects srvany's AppEnvironment parameter.

Since version 2.13, NSSM is translated into French.
Thanks François-Régis Tardy.

Since version 2.15, NSSM is translated into Italian.
Thanks Riccardo Gusmeroli.

Since version 2.17, NSSM can try to shut down console applications by
simulating a Control-C keypress.  If they have installed a handler routine
they can clean up and shut down gracefully on receipt of the event.

Usage
-----
In the usage notes below, arguments to the program may be written in angle 
brackets and/or square brackets.  <string> means you must insert the 
appropriate string and [<string>] means the string is optional.  See the 
examples below...


Installation using the GUI
--------------------------
To install a service, run

    nssm install <servicename>

You will be prompted to enter the full path to the application you wish 
to run and any command line options to pass to that application.

Use the system service manager (services.msc) to control advanced service 
properties such as startup method and desktop interaction.  NSSM may 
support these options at a later time...


Installation using the command line
-----------------------------------
To install a service, run

    nssm install <servicename> <application> [<options>]

NSSM will then attempt to install a service which runs the named application 
with the given options (if you specified any).

Don't forget to enclose paths in "quotes" if they contain spaces!


Managing the service
--------------------
NSSM will launch the application listed in the registry when you send it a 
start signal and will terminate it when you send a stop signal.  So far, so 
much like srvany.  But NSSM is the Non-Sucking service manager and can take 
action if/when the application dies.

With no configuration from you, NSSM will try to restart itself if it notices
that the application died but you didn't send it a stop signal.  NSSM will
keep trying, pausing between each attempt, until the service is successfully
started or you send it a stop signal.

NSSM will pause an increasingly longer time between subsequent restart attempts
if the service fails to start in a timely manner, up to a maximum of four
minutes.  This is so it does not consume an excessive amount of CPU time trying
to start a failed application over and over again.  If you identify the cause
of the failure and don't want to wait you can use the Windows service console
(where the service will be shown in Paused state) to send a continue signal to
NSSM and it will retry within a few seconds.

By default, NSSM defines "a timely manner" to be within 1500 milliseconds.
You can change the threshold for the service by setting the number of
milliseconds as a REG_DWORD value in the registry at
HKLM\SYSTEM\CurrentControlSet\Services\<service>\Parameters\AppThrottle.

NSSM will look in the registry under
HKLM\SYSTEM\CurrentControlSet\Services\<service>\Parameters\AppExit for
string (REG_EXPAND_SZ) values corresponding to the exit code of the application.
If the application exited with code 1, for instance, NSSM will look for a
string value under AppExit called "1" or, if it does not find it, will
fall back to the AppExit (Default) value.  You can find out the exit code
for the application by consulting the system event log.  NSSM will log the
exit code when the application exits.

Based on the data found in the registry, NSSM will take one of three actions:

If the value data is "Restart" NSSM will try to restart the application as
described above.  This is its default behaviour.

If the value data is "Ignore" NSSM will not try to restart the application
but will continue running itself.  This emulates the (usually undesirable)
behaviour of srvany.  The Windows Services console would show the service
as still running even though the application has exited.

If the value data is "Exit" NSSM will exit gracefully.  The Windows Services
console would show the service as stopped.  If you wish to provide
finer-grained control over service recovery you should use this code and
edit the failure action manually.  Please note that Windows versions prior
to Vista will not consider such an exit to be a failure.  On older versions
of Windows you should use "Suicide" instead.

If the value data is "Suicide" NSSM will simulate a crash and exit without
informing the service manager.  This option should only be used for
pre-Vista systems where you wish to apply a service recovery action.  Note
that if the monitored application exits with code 0, NSSM will only honour a
request to suicide if you explicitly configure a registry key for exit code 0.
If only the default action is set to Suicide NSSM will instead exit gracefully.


Removing services using the GUI
-------------------------------
NSSM can also remove services.  Run

    nssm remove <servicename>

to remove a service.  You will prompted for confirmation before the service 
is removed.  Try not to remove essential system services...


Removing service using the command line
---------------------------------------
To remove a service without confirmation from the GUI, run

    nssm remove <servicename> confirm

Try not to remove essential system services...


Logging
-------
NSSM logs to the Windows event log.  It registers itself as an event log source
and uses unique event IDs for each type of message it logs.  New versions may
add event types but existing event IDs will never be changed.

Because of the way NSSM registers itself you should be aware that you may not
be able to replace the NSSM binary if you have the event viewer open and that
running multiple instances of NSSM from different locations may be confusing if
they are not all the same version.


Example usage
-------------
To install an Unreal Tournament server:

    nssm install UT2004 c:\games\ut2004\system\ucc.exe server

To remove the server:

    nssm remove UT2004 confirm


Building NSSM from source
-------------------------
NSSM is known to compile with Visual Studio 6, Visual Studio 2005 and Visual
Studio 2008.

NSSM will also compile with Visual Studio 2010 but the resulting executable
will not run on versions of Windows older than XP SP2.


Credits
-------
Thanks to Iain Patterson for starting the NSSM project.
Thanks to Bernard Loh for finding a bug with service recovery.
Thanks to Benjamin Mayrargue (www.softlion.com) for adding 64-bit support.
Thanks to Joel Reingold for spotting a command line truncation bug.
Thanks to Arve Knudsen for spotting that child processes of the monitored
application could be left running on service shutdown, and that a missing
registry value for AppDirectory confused NSSM.
Thanks to Peter Wagemans and Laszlo Keresztfalvi for suggesting throttling restarts.
Thanks to Eugene Lifshitz for finding an edge case in CreateProcess() and for
advising how to build messages.mc correctly in paths containing spaces.
Thanks to Rob Sharp for pointing out that NSSM did not respect the
AppEnvironment registry value used by srvany.
Thanks to Szymon Nowak for help with Windows 2000 compatibility.
Thanks to François-Régis Tardy for French translation.
Thanks to Emilio Frini for spotting that French was inadvertently set as
the default language when the user's display language was not translated.
Thanks to Riccardo Gusmeroli for Italian translation.
Thanks to Eric Cheldelin for the inspiration to generate a Control-C event
on shutdown.

Licence
-------
NSSM is public domain.  You may unconditionally use it and/or its source code 
for any purpose you wish.


ChangeLog.txt
Changes since 2.16
-----------------
  * Allow building on Visual Studio Express.

  * Silently ignore INTERROGATE control.

  * Try to send Control-C events to console applications when
    shutting them down.

  * Fixed a handful of compile warnings.

Changes since 2.15
-----------------
  * Fixed case where NSSM could kill unrelated processes when
    shutting down.

Changes since 2.14
-----------------
  * NSSM is now translated into Italian.

  * Fixed GUI not allowing paths longer than 256 characters.

Changes since 2.13
-----------------
  * Fixed default GUI language being French not English.

Changes since 2.12
-----------------
  * Fixed failure to run on Windows 2000.

Changes since 2.11
-----------------
  * NSSM is now translated into French.

  * Really ensure systems recovery actions can happen.

    The change supposedly introduced in v2.4 to allow service recovery
    actions to be activated when the application exits gracefully with
    a non-zero error code didn't actually work.

Changes since 2.10
-----------------
  * Support AppEnvironment for compatibility with srvany.

Changes since 2.9
-----------------
  * Fixed failure to compile messages.mc in paths containing spaces.

  * Fixed edge case with CreateProcess().

    Correctly handle the case where the application executable is under
    a path which contains space and an executable sharing the initial
    part of that path (up to a space) exists.

Changes since 2.8
-----------------
  * Fixed failure to run on Windows versions prior to Vista.

Changes since 2.7
-----------------
  * Read Application, AppDirectory and AppParameters before each restart so
    a change to any one doesn't require restarting NSSM itself.

  * Fixed messages not being sent to the event log correctly in some
    cases.

  * Try to handle (strictly incorrect) quotes in AppDirectory.

    Windows directories aren't allowed to contain quotes so CreateProcess()
    will fail if the AppDirectory is quoted.  Note that it succeeds even if
    Application itself is quoted as the application plus parameters are
    interpreted as a command line.

  * Fixed failed to write full arguments to AppParameters when
    installing a service.

  * Throttle restarts.

    Back off from restarting the application immediately if it starts
    successfully but exits too soon.  The default value of "too soon" is
    1500 milliseconds.  This can be configured by adding a DWORD value
    AppThrottle to the registry.
    
    Handle resume messages from the service console to restart the
    application immediately even if it is throttled.

  * Try to kill the process tree gracefully.

    Before calling TerminateProcess() on all processes assocatiated with
    the monitored application, enumerate all windows and threads and
    post appropriate messages to them.  If the application bothers to
    listen for such messages it has a chance to shut itself down gracefully.

Changes since 2.6
-----------------
  * Handle missing registry values.

    Warn if AppParameters is missing.  Warn if AppDirectory is missing or
    unset and choose a fallback directory.
    First try to find the parent directory of the application.  If that
    fails, eg because the application path is just "notepad" or something,
    start in the Windows directory.

  * Kill process tree when stopping service.

    Ensure that all child processes of the monitored application are
    killed when the service stops by recursing through all running
    processes and terminating those whose parent is the application
    or one of its descendents.

Changes since 2.5
-----------------
  * Removed incorrect ExpandEnvironmentStrings() error.

    A log_event() call was inadvertently left in the code causing an error
    to be set to the eventlog saying that ExpandEnvironmentStrings() had
    failed when it had actually succeeded.

Changes since 2.4
-----------------
  * Allow use of REG_EXPAND_SZ values in the registry.

  * Don't suicide on exit status 0 by default.

    Suiciding when the application exits 0 will cause recovery actions to be
    taken.  Usually this is inappropriate.  Only suicide if there is an
    explicit AppExit value for 0 in the registry.
    
    Technically such behaviour could be abused to do something like run a
    script after successful completion of a service but in most cases a
    suicide is undesirable when no actual failure occurred.

  * Don't hang if startup parameters couldn't be determined.
    Instead, signal that the service entered the STOPPED state.
    Set START_PENDING state prior to actual startup.

Changes since 2.3
-----------------
  * Ensure systems recovery actions can happen.

    In Windows versions earlier than Vista the service manager would only
    consider a service failed (and hence eligible for recovery action) if
    the service exited without setting its state to SERVICE_STOPPED, even if
    it signalled an error exit code.
    In Vista and later the service manager can be configured to treat a
    graceful shutdown with error code as a failure but this is not the
    default behaviour.

    Try to configure the service manager to use the new behaviour when
    starting the service so users who set AppExit to Exit can use recovery
    actions as expected.

    Also recognise the new AppExit option Suicide for use on pre-Vista
    systems.  When AppExit is Suicide don't stop the service but exit
    inelegantly, which should be seen as a failure.

Changes since 2.2
-----------------
  * Send properly formatted messages to the event log.

  * Fixed truncation of very long path lengths in the registry.

Changes since 2.1
-----------------
 *  Decide how to handle application exit.

    When the service exits with exit code n look in
    HKLM\SYSTEM\CurrentControlSet\Services\<service>\Parameters\AppExit\<n>,
    falling back to the unnamed value if no such code is listed.  Parse the
    (string) value of this entry as follows:


gui.cpp
#include "nssm.h"

int nssm_gui(int resource, char *name) {
  /* Create window */
  HWND dlg = CreateDialog(0, MAKEINTRESOURCE(resource), 0, install_dlg);
  if (! dlg) {
    popup_message(MB_OK, NSSM_GUI_CREATEDIALOG_FAILED, error_string(GetLastError()));
    return 1;
  }

  /* Display the window */
  centre_window(dlg);
  ShowWindow(dlg, SW_SHOW);

  /* Set service name if given */
  if (name) {
    SetDlgItemText(dlg, IDC_NAME, name);
    /* No point making user click remove if the name is already entered */
    if (resource == IDD_REMOVE) {
      HWND button = GetDlgItem(dlg, IDC_REMOVE);
      if (button) {
        SendMessage(button, WM_LBUTTONDOWN, 0, 0);
        SendMessage(button, WM_LBUTTONUP, 0, 0);
      }
    }
  }

  /* Go! */
  MSG message;
  while (GetMessage(&message, 0, 0, 0)) {
    TranslateMessage(&message);
    DispatchMessage(&message);
  }

  return (int) message.wParam;
}

void centre_window(HWND window) {
  HWND desktop;
  RECT size, desktop_size;
  unsigned long x, y;

  if (! window) return;

  /* Find window size */
  if (! GetWindowRect(window, &size)) return;
  
  /* Find desktop window */
  desktop = GetDesktopWindow();
  if (! desktop) return;

  /* Find desktop window size */
  if (! GetWindowRect(desktop, &desktop_size)) return;

  /* Centre window */
  x = (desktop_size.right - size.right) / 2;
  y = (desktop_size.bottom - size.bottom) / 2;
  MoveWindow(window, x, y, size.right - size.left, size.bottom - size.top, 0);
}

/* Install the service */
int install(HWND window) {
  if (! window) return 1;

  /* Check parameters in the window */
  char name[VALUE_LENGTH];
  char exe[EXE_LENGTH];
  char flags[VALUE_LENGTH];

  /* Get service name */
  if (! GetDlgItemText(window, IDC_NAME, name, sizeof(name))) {
    popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_GUI_MISSING_SERVICE_NAME);
    return 2;
  }

  /* Get executable name */
  if (! GetDlgItemText(window, IDC_PATH, exe, sizeof(exe))) {
    popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_GUI_MISSING_PATH);
    return 3;
  }

  /* Get flags */
  if (SendMessage(GetDlgItem(window, IDC_FLAGS), WM_GETTEXTLENGTH, 0, 0)) {
    if (! GetDlgItemText(window, IDC_FLAGS, flags, sizeof(flags))) {
      popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_GUI_INVALID_OPTIONS);
      return 4;
    }
  }
  else ZeroMemory(&flags, sizeof(flags));

  /* See if it works */
  switch (install_service(name, exe, flags)) {
    case 2:
      popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_MESSAGE_OPEN_SERVICE_MANAGER_FAILED);
      return 2;

    case 3:
      popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_MESSAGE_PATH_TOO_LONG, NSSM);
      return 3;

    case 4:
      popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_GUI_OUT_OF_MEMORY_FOR_IMAGEPATH);
      return 4;

    case 5:
      popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_GUI_INSTALL_SERVICE_FAILED);
      return 5;

    case 6:
      popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_GUI_CREATE_PARAMETERS_FAILED);
      return 6;
  }

  popup_message(MB_OK, NSSM_MESSAGE_SERVICE_INSTALLED, name);
  return 0;
}

/* Remove the service */
int remove(HWND window) {
  if (! window) return 1;

  /* Check parameters in the window */
  char name[VALUE_LENGTH];

  /* Get service name */
  if (! GetDlgItemText(window, IDC_NAME, name, sizeof(name))) {
    popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_GUI_MISSING_SERVICE_NAME);
    return 2;
  }

  /* Confirm */
  if (popup_message(MB_YESNO, NSSM_GUI_ASK_REMOVE_SERVICE, name) != IDYES) return 0;

  /* See if it works */
  switch (remove_service(name)) {
    case 2:
      popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_MESSAGE_OPEN_SERVICE_MANAGER_FAILED);
      return 2;

    case 3:
      popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_GUI_SERVICE_NOT_INSTALLED);
      return 3;

    case 4:
      popup_message(MB_OK | MB_ICONEXCLAMATION, NSSM_GUI_REMOVE_SERVICE_FAILED);
      return 4;
  }

  popup_message(MB_OK, NSSM_MESSAGE_SERVICE_REMOVED, name);
  return 0;
}

/* Browse for application */
void browse(HWND window) {
  if (! window) return;

  unsigned long bufsize = 256;
  unsigned long len = bufsize;
  OPENFILENAME ofn;
  ZeroMemory(&ofn, sizeof(ofn));
  ofn.lStructSize = sizeof(ofn);
  ofn.lpstrFilter = (char *) HeapAlloc(GetProcessHeap(), 0, bufsize);
  /* XXX: Escaping nulls with FormatMessage is tricky */
  if (ofn.lpstrFilter) {
    ZeroMemory((void *) ofn.lpstrFilter, bufsize);
    char *localised = message_string(NSSM_GUI_BROWSE_FILTER_APPLICATIONS);
    _snprintf_s((char *) ofn.lpstrFilter, bufsize, _TRUNCATE, localised);
    /* "Applications" + NULL + "*.exe" + NULL */
    len = strlen(localised) + 1;
    LocalFree(localised);
    _snprintf_s((char *) ofn.lpstrFilter + len, bufsize - len, _TRUNCATE, "*.exe");
    /* "All files" + NULL + "*.*" + NULL */
    len += 6;
    localised = message_string(NSSM_GUI_BROWSE_FILTER_ALL_FILES);
    _snprintf_s((char *) ofn.lpstrFilter + len, bufsize - len, _TRUNCATE, localised);
    len += strlen(localised) + 1;
    LocalFree(localised);
    _snprintf_s((char *) ofn.lpstrFilter + len, bufsize - len, _TRUNCATE, "*.*");
    /* Remainder of the buffer is already zeroed */
  }
  ofn.lpstrFile = new char[MAX_PATH];
  ofn.lpstrFile[0] = '\0';
  ofn.lpstrTitle = message_string(NSSM_GUI_BROWSE_TITLE);
  ofn.nMaxFile = MAX_PATH;
  ofn.Flags = OFN_EXPLORER | OFN_PATHMUSTEXIST | OFN_HIDEREADONLY;

  if (GetOpenFileName(&ofn)) {
    SendMessage(window, WM_SETTEXT, 0, (LPARAM) ofn.lpstrFile);
  }
  if (ofn.lpstrFilter) HeapFree(GetProcessHeap(), 0, (void *) ofn.lpstrFilter);

  delete[] ofn.lpstrFile;
}

/* Install/remove dialogue callback */
INT_PTR CALLBACK install_dlg(HWND window, UINT message, WPARAM w, LPARAM l) {
  switch (message) {
    /* Creating the dialogue */
    case WM_INITDIALOG:
      return 1;

    /* Button was pressed or control was controlled */
    case WM_COMMAND:
      switch (LOWORD(w)) {
        /* OK button */
        case IDC_OK:
          PostQuitMessage(install(window));
          break;

        /* Cancel button */
        case IDC_CANCEL:
          DestroyWindow(window);
          break;

        /* Browse button */
        case IDC_BROWSE:
          browse(GetDlgItem(window, IDC_PATH));
          break;

        /* Remove button */
        case IDC_REMOVE:
          PostQuitMessage(remove(window));
          break;
      }
      return 1;

    /* Window closing */
    case WM_CLOSE:
      DestroyWindow(window);
      return 0;
    case WM_DESTROY:
      PostQuitMessage(0);
  }
  return 0;
}


process.cpp
#include "nssm.h"

int get_process_creation_time(HANDLE process_handle, FILETIME *ft) {
  FILETIME creation_time, exit_time, kernel_time, user_time;

  if (! GetProcessTimes(process_handle, &creation_time, &exit_time, &kernel_time, &user_time)) {
    log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_GETPROCESSTIMES_FAILED, error_string(GetLastError()), 0);
    return 1;
  }

  memmove(ft, &creation_time, sizeof(creation_time));

  return 0;
}

int get_process_exit_time(HANDLE process_handle, FILETIME *ft) {
  FILETIME creation_time, exit_time, kernel_time, user_time;

  if (! GetProcessTimes(process_handle, &creation_time, &exit_time, &kernel_time, &user_time)) {
    log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_GETPROCESSTIMES_FAILED, error_string(GetLastError()), 0);
    return 1;
  }

  memmove(ft, &exit_time, sizeof(exit_time));

  return 0;
}

int check_parent(char *service_name, PROCESSENTRY32 *pe, unsigned long ppid, FILETIME *pft, FILETIME *exit_time) {
  /* Check parent process ID matches. */
  if (pe->th32ParentProcessID != ppid) return 1;

  /*
    Process IDs can be reused so do a sanity check by making sure the child
    has been running for less time than the parent.
    Though unlikely, it's possible that the parent exited and its process ID
    was already reused, so we'll also compare against its exit time.
  */
  HANDLE process_handle = OpenProcess(PROCESS_QUERY_INFORMATION, false, pe->th32ProcessID);
  if (! process_handle) {
    char pid_string[16];
    _snprintf_s(pid_string, sizeof(pid_string), "%d", _TRUNCATE, pe->th32ProcessID);
    log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_OPENPROCESS_FAILED, pid_string, service_name, error_string(GetLastError()), 0);
    return 2;
  }

  FILETIME ft;
  if (get_process_creation_time(process_handle, &ft)) {
    CloseHandle(process_handle);
    return 3;
  }

  CloseHandle(process_handle);

  /* Verify that the parent's creation time is not later. */
  if (CompareFileTime(pft, &ft) > 0) return 4;

  /* Verify that the parent's exit time is not earlier. */
  if (CompareFileTime(exit_time, &ft) < 0) return 5;

  return 0;
}

/* Send some window messages and hope the window respects one or more. */
int CALLBACK kill_window(HWND window, LPARAM arg) {
  kill_t *k = (kill_t *) arg;

  unsigned long pid;
  if (! GetWindowThreadProcessId(window, &pid)) return 1;
  if (pid != k->pid) return 1;

  /* First try sending WM_CLOSE to request that the window close. */
  k->signalled |= PostMessage(window, WM_CLOSE, k->exitcode, 0);

  /*
    Then tell the window that the user is logging off and it should exit
    without worrying about saving any data.
  */
  k->signalled |= PostMessage(window, WM_ENDSESSION, 1, ENDSESSION_CLOSEAPP | ENDSESSION_CRITICAL | ENDSESSION_LOGOFF);

  return 1;
}

/*
  Try to post a message to the message queues of threads associated with the
  given process ID.  Not all threads have message queues so there's no
  guarantee of success, and we don't want to be left waiting for unsignalled
  processes so this function returns only true if at least one thread was
  successfully prodded.
*/
int kill_threads(char *service_name, kill_t *k) {
  int ret = 0;

  /* Get a snapshot of all threads in the system. */
  HANDLE snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPTHREAD, 0);
  if (! snapshot) {
    log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_CREATETOOLHELP32SNAPSHOT_THREAD_FAILED, service_name, error_string(GetLastError()), 0);
    return 0;
  }

  THREADENTRY32 te;
  ZeroMemory(&te, sizeof(te));
  te.dwSize = sizeof(te);

  if (! Thread32First(snapshot, &te)) {
    log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_THREAD_ENUMERATE_FAILED, service_name, error_string(GetLastError()), 0);
    CloseHandle(snapshot);
    return 0;
  }

  /* This thread belongs to the doomed process so signal it. */
  if (te.th32OwnerProcessID == k->pid) {
    ret |= PostThreadMessage(te.th32ThreadID, WM_QUIT, k->exitcode, 0);
  }

  while (true) {
    /* Try to get the next thread. */
    if (! Thread32Next(snapshot, &te)) {
      unsigned long error = GetLastError();
      if (error == ERROR_NO_MORE_FILES) break;
      log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_THREAD_ENUMERATE_FAILED, service_name, error_string(GetLastError()), 0);
      CloseHandle(snapshot);
      return ret;
    }

    if (te.th32OwnerProcessID == k->pid) {
      ret |= PostThreadMessage(te.th32ThreadID, WM_QUIT, k->exitcode, 0);
    }
  }

  CloseHandle(snapshot);

  return ret;
}

/* Give the process a chance to die gracefully. */
int kill_process(char *service_name, HANDLE process_handle, unsigned long pid, unsigned long exitcode) {
  /* Shouldn't happen. */
  if (! pid) return 1;
  if (! process_handle) return 1;

  unsigned long ret;
  if (GetExitCodeProcess(process_handle, &ret)) {
    if (ret != STILL_ACTIVE) return 1;
  }

  kill_t k = { pid, exitcode, 0 };

  /* Try to send a Control-C event to the console. */
  if (! kill_console(service_name, process_handle, pid)) return 1;

  /*
    Try to post messages to the windows belonging to the given process ID.
    If the process is a console application it won't have any windows so there's
    no guarantee of success.
  */
  EnumWindows((WNDENUMPROC) kill_window, (LPARAM) &k);
  if (k.signalled) {
    if (! WaitForSingleObject(process_handle, NSSM_KILL_WINDOW_GRACE_PERIOD)) return 1;
  }

  /*
    Try to post messages to any thread message queues associated with the
    process.  Console applications might have them (but probably won't) so
    there's still no guarantee of success.
  */
  if (kill_threads(service_name, &k)) {
    if (! WaitForSingleObject(process_handle, NSSM_KILL_THREADS_GRACE_PERIOD)) return 1;
  }

  /* We tried being nice.  Time for extreme prejudice. */
  return TerminateProcess(process_handle, exitcode);
}

/* Simulate a Control-C event to our console (shared with the app). */
int kill_console(char *service_name, HANDLE process_handle, unsigned long pid) {
  unsigned long ret;

  /* Try to attach to the process's console. */
  if (! AttachConsole(pid)) {
    ret = GetLastError();

    switch (ret) {
      case ERROR_INVALID_HANDLE:
        /* The app doesn't have a console. */
        return 1;

      case ERROR_GEN_FAILURE:
        /* The app already exited. */
        return 2;

      case ERROR_ACCESS_DENIED:
      default:
        /* We already have a console. */
        log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_ATTACHCONSOLE_FAILED, service_name, error_string(ret), 0);
        return 3;
    }
  }

  /* Ignore the event ourselves. */
  ret = 0;
  if (! SetConsoleCtrlHandler(0, TRUE)) {
    log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_SETCONSOLECTRLHANDLER_FAILED, service_name, error_string(GetLastError()), 0);
    ret = 4;
  }

  /* Sent the event. */
  if (! ret) {
    if (! GenerateConsoleCtrlEvent(CTRL_C_EVENT, 0)) {
      log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_GENERATECONSOLECTRLEVENT_FAILED, service_name, error_string(GetLastError()), 0);
      ret = 5;
    }
  }

  /* Detach from the console. */
  if (! FreeConsole()) {
    log_event(EVENTLOG_WARNING_TYPE, NSSM_EVENT_FREECONSOLE_FAILED, service_name, error_string(GetLastError()), 0);
  }

  /* Wait for process to exit. */
  if (! WaitForSingleObject(process_handle, NSSM_KILL_CONSOLE_GRACE_PERIOD)) return 6;

  return ret;
}

void kill_process_tree(char *service_name, unsigned long pid, unsigned long exitcode, unsigned long ppid, FILETIME *parent_creation_time, FILETIME *parent_exit_time) {
  /* Shouldn't happen unless the service failed to start. */
  if (! pid) return;

  char pid_string[16], code[16];
  _snprintf_s(pid_string, sizeof(pid_string), _TRUNCATE, "%d", pid);
  _snprintf_s(code, sizeof(code), "%d", _TRUNCATE, exitcode);
  log_event(EVENTLOG_INFORMATION_TYPE, NSSM_EVENT_KILLING, service_name, pid_string, code, 0);

  /* Get a snapshot of all processes in the system. */
  HANDLE snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
  if (! snapshot) {
    log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_CREATETOOLHELP32SNAPSHOT_PROCESS_FAILED, service_name, error_string(GetLastError()), 0);
    return;
  }

  PROCESSENTRY32 pe;
  ZeroMemory(&pe, sizeof(pe));
  pe.dwSize = sizeof(pe);

  if (! Process32First(snapshot, &pe)) {
    log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_PROCESS_ENUMERATE_FAILED, service_name, error_string(GetLastError()), 0);
    CloseHandle(snapshot);
    return;
  }

  /* This is a child of the doomed process so kill it. */
  if (! check_parent(service_name, &pe, pid, parent_creation_time, parent_exit_time)) kill_process_tree(service_name, pe.th32ProcessID, exitcode, ppid, parent_creation_time, parent_exit_time);

  while (true) {
    /* Try to get the next process. */
    if (! Process32Next(snapshot, &pe)) {
      unsigned long ret = GetLastError();
      if (ret == ERROR_NO_MORE_FILES) break;
      log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_PROCESS_ENUMERATE_FAILED, service_name, error_string(GetLastError()), 0);
      CloseHandle(snapshot);
      return;
    }

    if (! check_parent(service_name, &pe, pid, parent_creation_time, parent_exit_time)) kill_process_tree(service_name, pe.th32ProcessID, exitcode, ppid, parent_creation_time, parent_exit_time);
  }

  CloseHandle(snapshot);

  /* We will need a process handle in order to call TerminateProcess() later. */
  HANDLE process_handle = OpenProcess(SYNCHRONIZE | PROCESS_QUERY_INFORMATION | PROCESS_VM_READ | PROCESS_TERMINATE, false, pid);
  if (! process_handle) {
    log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_OPENPROCESS_FAILED, pid_string, service_name, error_string(GetLastError()), 0);
    return;
  }

  char ppid_string[16];
  _snprintf_s(ppid_string, sizeof(ppid_string), _TRUNCATE, "%d", ppid);
  log_event(EVENTLOG_INFORMATION_TYPE, NSSM_EVENT_KILL_PROCESS_TREE, pid_string, ppid_string, service_name, 0);
  if (! kill_process(service_name, process_handle, pid, exitcode)) {
    /* Maybe it already died. */
    unsigned long ret;
    if (! GetExitCodeProcess(process_handle, &ret) || ret == STILL_ACTIVE) log_event(EVENTLOG_ERROR_TYPE, NSSM_EVENT_TERMINATEPROCESS_FAILED, pid_string, service_name, error_string(GetLastError()), 0);
  }

  CloseHandle(process_handle);
}


nssm.h
#ifndef NSSM_H
#define NSSM_H

#define _WIN32_WINNT 0x0500
#include <shlwapi.h>
#include <stdarg.h>
#include <stdio.h>
#include <windows.h>
#include "event.h"
#include "messages.h"
#include "process.h"
#include "registry.h"
#include "service.h"
#include "gui.h"

int str_equiv(const char *, const char *);

#define NSSM "nssm"
#define NSSM_VERSION "2.17-dev"
#define NSSM_DATE "2013-08-17"

/*
  MSDN says the commandline in CreateProcess() is limited to 32768 characters
  and the application name to MAX_PATH.
  A registry key is limited to 255 characters.
  A registry value is limited to 16383 characters.
  Therefore we limit the service name to accommodate the path under HKLM.
*/
#define EXE_LENGTH MAX_PATH
#define CMD_LENGTH 32768
#define KEY_LENGTH 255
#define VALUE_LENGTH 16383
#define SERVICE_NAME_LENGTH KEY_LENGTH - 55

/*
  Throttle the restart of the service if it stops before this many
  milliseconds have elapsed since startup.  Override in registry.
*/
#define NSSM_RESET_THROTTLE_RESTART 1500

/*
  How many milliseconds to wait for the application to die after sending
  a Control-C event to its console.
*/
#define NSSM_KILL_CONSOLE_GRACE_PERIOD 1500
/*
  How many milliseconds to wait for the application to die after posting to
  its windows' message queues.
*/
#define NSSM_KILL_WINDOW_GRACE_PERIOD 1500
/*
  How many milliseconds to wait for the application to die after posting to
  its threads' message queues.
*/
#define NSSM_KILL_THREADS_GRACE_PERIOD 1500

/* Margin of error for service status wait hints in milliseconds. */
#define NSSM_WAITHINT_MARGIN 2000

#endif
